### ISSUE TITLE
All bytes of instruction fetch matching the same PMP entry?

### ISSUE THREAD

**Issue by ma142857** [riscv/riscv-isa-manual#880]
https://github.com/riscv/riscv-isa-manual/blob/190ed91590f6f58f3ac3236a8c7081aec1643ae0/src/machine.tex#L3757
The Privileged spec states that:
_"The matching PMP entry must match all bytes of an access, or the access fails, irrespective of the L, R, W, and X bits."_
For instruction fetch, does it mean that every bytes of the physical access related to the instruction fetch shall all match the same PMP entry?
Due to C-ext, 32-bit instruction could across virtual page boundary, and lead to non-continuous physical accesses for a single instruction fetch. Is it required that the 2 non-continuous physical memory accesses related to the instruction fetch have to match the same PMP entry to pass PMP check?

**Comment by allenjbaum** [comment]
As I understand it... it's complicated. If the access goes out as a
*single* transaction, then yes.
But, if it crosses the boundary and is broken up into *separate*
transactions (which would be typical),
then each of them individually needs to pass PMP checks - but each could be
a different PMP entry..

**Comment by ma142857** [comment]
Thanks @allenjbaum for the quick response!
As you state, an **_access_** could be broken up into more than one **_transactions_**.  Personally I don't foresee a problem if e.g. all the transactions related to an access pass PMP but falls in different PMP entries. But the statement in the privileged spec is definitely confusing since an _**access**_  usually means the whole fetch/load/store access itself architecturally (instead of broken up transactions).  Could you update the privileged spec to clarify this?
BTW, I think that also align with Spike implementation.

**Comment by scottj97** [comment]
This is already there in the PMP section of the privileged spec:

> Failed accesses generate an instruction, load, or store access-fault exception. Note that a single instruction may generate multiple accesses, which may not be mutually atomic. An access-fault exception is generated if at least one access generated by an instruction fails, though other accesses generated by that instruction may succeed with visible side effects. Notably, instructions that reference virtual memory are decomposed into multiple accesses.

> On some implementations, misaligned loads, stores, and instruction fetches may also be decomposed into multiple accesses, some of which may succeed before an access-fault exception occurs.

The privileged spec does not use the term "transaction". One instruction (e.g. `SD`) can be broken up into multiple accesses (if misaligned), including two for the instruction fetch and two more for the store.

**Comment by aswaterman** [comment]
I don’t believe any spec updates are necessary, as the following paragraph covers this case: https://github.com/riscv/riscv-isa-manual/blob/master/src/machine.tex#L3785

**Comment by ma142857** [comment]
> Note that a single instruction may generate multiple accesses, which may not be mutually atomic.
I see, the following sections further defines what does _**access**_ means for the previous section. 
Usually the term should be defined first then referred...
But I agree, it is okay to not update the spec. Thanks everyone for the clarification.

**Comment by aswaterman** [comment]
Closed as answered.
